### ISSUE TITLE
Convert sim-facing code to TypeScript

### ISSUE THREAD

**Issue by samreid** [phetsims/joist#776]
Such as Screen, ScreenView.  This came up during https://github.com/phetsims/center-and-spread/issues/6

**Comment by samreid** [comment]
Screen.js was converted to TypeScript.  For this change, it was also made to be a parametric type.  I updated existing TypeScript screens.

**Comment by samreid** [comment]
I also converted ScreenView to TypeScript.

**Comment by samreid** [comment]
I converted Screen and ScreenView to TypeScript.  A code review would be appropriate.  @pixelzoom or @jbphet do you have time?

**Comment by pixelzoom** [comment]
Screen.ts and ScreenView.ts look good.  I removed one unnecessary `assert` in the above commit.

I'm not convinced that the options API needs to be this wide for ScreenView:

```js
export type ScreenViewOptions = ScreenViewSelfOptions & NodeOptions;
```

... but I guess it's a reasonable starting point.

**Comment by pixelzoom** [comment]
@samreid I see that you parameterized SimulaRasaScreen and SimulaRasaScreenView, but did not convert `providedOptions` types, or change `merge` to `optionize`.   I'm still trying to wrap my head around the (complicated) `optionize` syntax.   Should the `providedOptions` and `optionize` look like this? 

```ts
// SimulaRasaScreen
  constructor( providedOptions: ScreenOptions ) {
    const options = optionize<ScreenOptions, {}, ScreenOptions, 'tandem'>( {

// SimulaRasaScreenView
  constructor( model: SimulaRasaModel, providedOptions: ScreenViewOptions ) {
    const options = optionize<ScreenViewOptions, {}, ScreenViewOptions, 'tandem'>( {
```

**Comment by pixelzoom** [comment]
`Sim` might be nice to convert next.  I found myself having to do this in Geometric Optics, and I suspect others are having to work around this too.

```js
//TODO https://github.com/phetsims/geometric-optics/issues/326 should be defined by Sim
type SimOptions = {
  credits?: object,
  hasKeyboardHelpContent?: boolean,
  createOptionsDialogContent?: ( tandem: Tandem ) => Node
}

...
const options = optionize<GOSimOptions, SelfOptions, SimOptions>( {
```

**Comment by samreid** [comment]
I'm interested in working on Sim.js => TypeScript.  I'll take a look.

**Comment by samreid** [comment]
I committed `ts-nocheck` and Sim.js => ts before I knew the scope of this project. I have a working prototype and would like to commit.  It is not complete, but it is a good start.  However, I won't be working later tonight to check when the CT column is ready.  But I will have time to address CT issues 8am-9am on April 13.  But if I wait to push until April 13 morning, then I'll be in meetings when the columns come out.  I feel it has been rare for developers to work much before 8am, so perhaps that would be reasonable.

But also, this is a commit that would be nice to have someone review soon.  So maybe I should continue in a branch?  I'm not sure.

Anyways, here's a patch for now.

<details>

```diff

Index: main/chipper/phet-types.d.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/chipper/phet-types.d.ts b/main/chipper/phet-types.d.ts
--- a/main/chipper/phet-types.d.ts	(revision 693e88bf3a398fcb3870c579abbfb2b7c6411530)
+++ b/main/chipper/phet-types.d.ts	(date 1649825816826)
@@ -15,4 +15,11 @@
 declare var QueryStringMachine: { // eslint-disable-line no-unused-vars, no-var
   getAll: ( a: any ) => any;
   containsKey: ( key: string ) => boolean;
-};
\ No newline at end of file
+  warnings: string[];
+};
+
+// globals used in Sim.ts
+declare var phetSplashScreenDownloadComplete: () => void; // eslint-disable-line no-unused-vars, no-var
+declare var TWEEN: { update: ( dt: number ) => void };  // eslint-disable-line no-unused-vars, no-var
+declare var phetSplashScreen: { dispose: () => void };  // eslint-disable-line no-unused-vars, no-var
+declare var phetio: any; // eslint-disable-line no-unused-vars, no-var
\ No newline at end of file
Index: main/joist/js/Sim.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/joist/js/Sim.ts b/main/joist/js/Sim.ts
--- a/main/joist/js/Sim.ts	(revision e990d9da4f1f89581564ef6192b6466e6e433fcf)
+++ b/main/joist/js/Sim.ts	(date 1649825320975)
@@ -1,10 +1,13 @@
 // Copyright 2013-2022, University of Colorado Boulder
-// @ts-nocheck
+
 /**
  * Main class that represents one simulation.
  * Provides default initialization, such as polyfills as well.
  * If the simulation has only one screen, then there is no homescreen, home icon or screen icon in the navigation bar.
  *
+ * The type for the contained Screen instances is Screen<any,any> since we do not want to parameterize Sim<[{M1,V1},{M2,V2}]
+ * etc.
+ *
  * @author Sam Reid (PhET Interactive Simulations)
  * @author Chris Malley (PixelZoom, Inc.)
  * @author Jonathan Olson <jonathan.olson@colorado.edu>
@@ -12,7 +15,7 @@
 
 import animationFrameTimer from '../../axon/js/animationFrameTimer.js';
 import BooleanProperty from '../../axon/js/BooleanProperty.js';
-import createObservableArray from '../../axon/js/createObservableArray.js';
+import createObservableArray, { ObservableArray } from '../../axon/js/createObservableArray.js';
 import DerivedProperty from '../../axon/js/DerivedProperty.js';
 import Emitter from '../../axon/js/Emitter.js';
 import NumberProperty from '../../axon/js/NumberProperty.js';
@@ -25,13 +28,14 @@
 import DotUtils from '../../dot/js/Utils.js'; // eslint-disable-line default-import-match-filename
 import merge from '../../phet-core/js/merge.js';
 import platform from '../../phet-core/js/platform.js';
+import optionize from '../../phet-core/js/optionize.js';
 import StringUtils from '../../phetcommon/js/util/StringUtils.js';
 import BarrierRectangle from '../../scenery-phet/js/BarrierRectangle.js';
 import { animatedPanZoomSingleton, globalKeyStateTracker, Node, Utils, voicingManager, voicingUtteranceQueue } from '../../scenery/js/imports.js';
 import '../../sherpa/lib/game-up-camera-1.0.0.js';
 import soundManager from '../../tambo/js/soundManager.js';
 import PhetioAction from '../../tandem/js/PhetioAction.js';
-import PhetioObject from '../../tandem/js/PhetioObject.js';
+import PhetioObject, { PhetioObjectOptions } from '../../tandem/js/PhetioObject.js';
 import Tandem from '../../tandem/js/Tandem.js';
 import NumberIO from '../../tandem/js/types/NumberIO.js';
 import audioManager from './audioManager.js';
@@ -57,6 +61,9 @@
 import LegendsOfLearningSupport from './thirdPartySupport/LegendsOfLearningSupport.js';
 import Toolbar from './toolbar/Toolbar.js';
 import updateCheck from './updateCheck.js';
+import PreferencesConfiguration from './preferences/PreferencesConfiguration.js';
+import IReadOnlyProperty from '../../axon/js/IReadOnlyProperty.js';
+import IntentionalAny from '../../phet-core/js/types/IntentionalAny.js';
 
 // constants
 const PROGRESS_BAR_WIDTH = 273;
@@ -74,25 +81,178 @@
 // mode in which it will only update the model + view based on the playback clock events rather than the system clock.
 // This must be set before the simulation is launched in order to ensure that no errant stepSimulation steps are called
 // before the playback events begin.  This value is overridden for playback by PhetioEngineIO.
-// @public (phet-io)
+// (phet-io)
 phet.joist.playbackModeEnabledProperty = new BooleanProperty( phet.chipper.queryParameters.playbackMode );
 
 assert && assert( typeof phet.chipper.brand === 'string', 'phet.chipper.brand is required to run a sim' );
 
-class Sim extends PhetioObject {
+type SelfOptions = {
+
+  // credits, see AboutDialog for format.  TODO https://github.com/phetsims/joist/issues/776 convert credits to TS and give a type
+  credits?: any;
+
+  // {null|function(tandem:Tandem):Node} creates the content for the Options dialog
+  createOptionsDialogContent?: null | ( ( t: Tandem ) => Node );
+
+  // a {Node} placed onto the home screen (if available)
+  homeScreenWarningNode?: null | Node;
+
+  // {boolean} - true when this sim supports a keyboard help button on the navigation bar that shows keyboard help
+  // content. This content is specific to each screen, see Screen.keyboardHelpNode for more info.
+  hasKeyboardHelpContent?: boolean;
+
+  // {PreferencesConfiguration|null} - If a PreferencesConfiguration is provided the sim will
+  // include the PreferencesDialog and a button in the NavigationBar to open it.
+  preferencesConfiguration?: PreferencesConfiguration | null;
+
+  // Passed to SimDisplay, but a top level option for API ease.
+  webgl?: boolean;
+
+  // Passed to the SimDisplay
+  // TODO: https://github.com/phetsims/joist/issues/776 convert SimDisplay to TypeScript
+  simDisplayOptions?: any;
+};
+
+export type SimOptions = SelfOptions & PhetioObjectOptions;
+
+export default class Sim extends PhetioObject {
+
+  // TODO: https://github.com/phetsims/joist/issues/776 try marking things as private/readonly
+  readonly simNameProperty: StringProperty;
+  readonly createOptionsDialogContent: ( ( t: Tandem ) => Node ) | null;
+
+  // Indicates sim construction completed, and that all screen models and views have been created.
+  // This was added for PhET-iO but can be used by any client. This does not coincide with the end of the Sim
+  // constructor (because Sim has asynchronous steps that finish after the constructor is completed)
+  readonly isConstructionCompleteProperty: Property<boolean>;
+
+  // Stores the effective window dimensions that the simulation will be taking up
+  readonly dimensionProperty: Property<Dimension2>;
+
+  // Indicates when the sim resized.  This Action is implemented so it can be automatically played back.
+  readonly resizeAction: PhetioAction<[ number, number ]>;
+
+  // (joist-internal)
+  readonly navigationBar: NavigationBar;
+  readonly homeScreen: HomeScreen | null;
+
+  // Indicates when a frame starts.  Listen to this Emitter if you have an action that must be
+  // performed before the step begins.
+  frameStartedEmitter: Emitter;
+
+  // Indicates when a frame ends.  Listen to this Emitter if you have an action that must be
+  // performed after the step completes.
+  frameEndedEmitter: Emitter;
+
+  // Steps the simulation. This Action is implemented so it can be automatically
+  // played back for PhET-iO record/playback.  Listen to this Action if you have an action that happens during the
+  // simulation step.
+  private readonly stepSimulationAction: PhetioAction<[ number ]>;
+
+  // the ordered list of sim-specific screens that appear in this runtime of the sim
+  readonly simScreens: Screen<IntentionalAny, IntentionalAny>[];
+
+  // all screens that appear in the runtime of this sim, with the homeScreen first if it was created
+  readonly screens: Screen<IntentionalAny, IntentionalAny>[];
+
+  // the displayed name in the sim. This depends on what screens are shown this runtime (effected by query parameters).
+  displayedSimNameProperty: IReadOnlyProperty<string>;
+  screenProperty: Property<Screen<IntentionalAny, IntentionalAny>>;
+
+  // true if all possible screens are present (order-independent)
+  readonly allScreensCreated: boolean;
+
+  // When the sim is active, scenery processes inputs and stepSimulation(dt) runs from the system clock.
+  // Set to false for when the sim will be paused.
+  readonly activeProperty: BooleanProperty;
+
+  // indicates whether the browser tab containing the simulation is currently visible
+  readonly browserTabVisibleProperty: BooleanProperty;
+
+  // (joist-internal) - How the home screen and navbar are scaled. This scale is based on the
+  // HomeScreen's layout bounds to support a consistently sized nav bar and menu. If this scale was based on the
+  // layout bounds of the current screen, there could be differences in the nav bar across screens.
+  readonly scaleProperty: NumberProperty;
+
+  // (joist-internal) global bounds for the entire simulation. null before first resize
+  boundsProperty: Property<Bounds2 | null>;
+
+  // (joist-internal) global bounds for the screen-specific part (excludes the navigation bar), null before first resize
+  screenBoundsProperty: Property<Bounds2 | null>;
+
+  lookAndFeel: LookAndFeel;
+  private destroyed: boolean;
+  readonly memoryMonitor: MemoryMonitor;
+
+  // public (read-only) {boolean} - if true, add support specific to accessible technology that work with touch devices.
+  supportsGestureDescription: boolean;
+  hasKeyboardHelpContent: boolean;
+
+  // if PhET-iO is currently setting the state of the simulation. See PhetioStateEngine for details. This must be
+  // declared before soundManager.initialized is called.
+  readonly isSettingPhetioStateProperty: IReadOnlyProperty<boolean>;
+
+  // (joist-internal)
+  version: string;
+
+  // (joist-internal)
+  credits: any;
+
+  // number of animation frames that have occurred
+  frameCounter: number;
+
+  // Whether the window has resized since our last updateDisplay()
+  resizePending: boolean;
+
+  // Make our locale available
+  locale: string;
+
+  // create this only after all other members have been set on Sim
+  readonly simInfo: SimInfo;
+  private readonly display: SimDisplay;
+
+  // The Toolbar is not created unless requested with a PreferencesConfiguration.
+  toolbar: Toolbar | null;
+
+  // If Preferences are available through a PreferencesConfiguration,
+  // this type will be added to the Sim to manage the state of features that can be enabled/disabled
+  // through user preferences.
+  preferencesManager: PreferencesManager | null;
+
+  // layer for popups, dialogs, and their backgrounds and barriers
+  topLayer: Node;
+
+  // root node for the Display
+  rootNode: Node;
+
+  // list of nodes that are "modal" and hence block input with the barrierRectangle.  Used by modal dialogs
+  // and the PhetMenu
+  private modalNodeStack: ObservableArray<Node>;
+
+  // (joist-internal) Semi-transparent black barrier used to block input events when a dialog (or other popup)
+  // is present, and fade out the background.
+  barrierRectangle: BarrierRectangle;
+
+  // Keep track of the previous time for computing dt, and initially signify that time hasn't been recorded yet.
+  lastStepTime: number;
+  lastAnimationFrameTime: number;
+
+  // (joist-internal) Bind the animation loop so it can be called from requestAnimationFrame with the right this.
+  boundRunAnimationLoop: () => void;
+  updateBackground: () => void;
 
   /**
-   * @param {string} name - the name of the simulation, to be displayed in the navbar and homescreen
-   * @param {Screen[]} allSimScreens - the possible screens for the sim in order of declaration (does not include the home screen)
-   * @param {Object} [options] - see below for options
+   * @param name - the name of the simulation, to be displayed in the navbar and homescreen
+   * @param allSimScreens - the possible screens for the sim in order of declaration (does not include the home screen)
+   * @param [providedOptions] - see below for options
    */
-  constructor( name, allSimScreens, options ) {
+  constructor( name: string, allSimScreens: Screen<IntentionalAny, IntentionalAny>[], providedOptions?: SimOptions ) {
 
     window.phetSplashScreenDownloadComplete();
 
     assert && assert( allSimScreens.length >= 1, 'at least one screen is required' );
 
-    options = merge( {
+    const options = optionize<SimOptions, SelfOptions, PhetioObjectOptions>( {
 
       // credits, see AboutDialog for format
       credits: {},
@@ -121,7 +281,7 @@
       phetioState: false,
       phetioReadOnly: true,
       tandem: Tandem.ROOT
-    }, options );
+    }, providedOptions );
 
     assert && assert( !options.simDisplayOptions.webgl, 'use top level sim option instead of simDisplayOptions' );
 
@@ -134,10 +294,9 @@
 
     super( options );
 
-    // @public (read-only PhetMenu) {null|function(tandem:Tandem):Node}
     this.createOptionsDialogContent = options.createOptionsDialogContent;
 
-    // @public {Property.<string>} (joist-internal)
+    // (joist-internal)
     this.simNameProperty = new StringProperty( name, {
       tandem: Tandem.GENERAL_MODEL.createTandem( 'simNameProperty' ),
       phetioFeatured: true,
@@ -151,21 +310,16 @@
       throw new Error( 'playbackModeEnabledProperty cannot be changed after Sim construction has begun' );
     } );
 
-    // @public {Property} that indicates sim construction completed, and that all screen models and views have been created.
-    // This was added for PhET-iO but can be used by any client. This does not coincide with the end of the Sim
-    // constructor (because Sim has asynchronous steps that finish after the constructor is completed)
-    this.isConstructionCompleteProperty = new Property( false );
+    this.isConstructionCompleteProperty = new Property<boolean>( false );
     assert && this.isConstructionCompleteProperty.lazyLink( isConstructionComplete => {
       assert && assert( isConstructionComplete, 'Sim construction should never uncomplete' );
     } );
 
-    // @public {Property.<Dimension2>} - Stores the effective window dimensions that the simulation will be taking up
     this.dimensionProperty = new Property( new Dimension2( 0, 0 ), {
       useDeepEquality: true
     } );
 
-    // @public - Action that indicates when the sim resized.  This Action is implemented so it can be automatically played back.
-    this.resizeAction = new PhetioAction( ( width, height ) => {
+    this.resizeAction = new PhetioAction<[ number, number ]>( ( width, height ) => {
       assert && assert( width > 0 && height > 0, 'sim should have a nonzero area' );
 
       this.dimensionProperty.value = new Dimension2( width, height );
@@ -195,6 +349,8 @@
 
       // Layout each of the screens
       _.each( this.screens, m => m.view.layout( availableScreenBounds ) );
+
+      // @ts-ignore
       this.topLayer.children.forEach( child => child.layout && child.layout( availableScreenBounds ) );
 
       // Fixes problems where the div would be way off center on iOS7
@@ -209,14 +365,14 @@
 
       // set the scale describing the target Node, since scale from window resize is applied to each ScreenView,
       // (children of the PanZoomListener targetNode)
-      animatedPanZoomSingleton.listener.setTargetScale( scale );
+      animatedPanZoomSingleton.listener!.setTargetScale( scale );
 
       // set the bounds which accurately describe the panZoomListener targetNode, since it would otherwise be
       // inaccurate with the very large BarrierRectangle
-      animatedPanZoomSingleton.listener.setTargetBounds( this.boundsProperty.value );
+      animatedPanZoomSingleton.listener!.setTargetBounds( this.boundsProperty.value );
 
       // constrain the simulation pan bounds so that it cannot be moved off screen
-      animatedPanZoomSingleton.listener.setPanBounds( this.boundsProperty.value );
+      animatedPanZoomSingleton.listener!.setPanBounds( this.boundsProperty.value );
     }, {
       tandem: Tandem.GENERAL_MODEL.createTandem( 'resizeAction' ),
       parameters: [
@@ -239,27 +395,19 @@
     // relatively awkward, it is possible to listen for a callback when a frame begins, when a frame is being processed
     // or after the frame is complete.  See https://github.com/phetsims/joist/issues/534
 
-    // @public Emitter that indicates when a frame starts.  Listen to this Emitter if you have an action that must be
-    // performed before the step begins.
     this.frameStartedEmitter = new Emitter();
-
-    // @public Emitter that indicates when a frame ends.  Listen to this Emitter if you have an action that must be
-    // performed after the step completes.
     this.frameEndedEmitter = new Emitter( {
       tandem: Tandem.GENERAL_MODEL.createTandem( 'frameEndedEmitter' ),
       phetioHighFrequency: true
     } );
 
-    // @public {Action} Action that steps the simulation. This Action is implemented so it can be automatically
-    // played back for PhET-iO record/playback.  Listen to this Action if you have an action that happens during the
-    // simulation step.
     this.stepSimulationAction = new PhetioAction( dt => {
       this.frameStartedEmitter.emit();
 
       // increment this before we can have an exception thrown, to see if we are missing frames
       this.frameCounter++;
 
-      // Apply any scaling effects here before it is used.
+      // Apply time scale effects here before usage
       dt *= phet.chipper.queryParameters.speed;
 
       if ( this.resizePending ) {
@@ -271,9 +419,7 @@
       const screen = this.screenProperty.value;
 
       // cap dt based on the current screen, see https://github.com/phetsims/joist/issues/130
-      if ( screen.maxDT ) {
-        dt = Math.min( dt, screen.maxDT );
-      }
+      dt = Math.min( dt, screen.maxDT );
 
       // TODO: we are /1000 just to *1000?  Seems wasteful and like opportunity for error. See https://github.com/phetsims/joist/issues/387
       // Store the elapsed time in milliseconds for usage by Tween clients
@@ -288,7 +434,7 @@
         screen.model.step( dt );
       }
 
-      // If using the TWEEN animation library, then update all of the tweens (if any) before rendering the scene.
+      // If using the TWEEN animation library, then update tweens before rendering the scene.
       // Update the tweens after the model is updated but before the view step.
       // See https://github.com/phetsims/joist/issues/401.
       //TODO https://github.com/phetsims/joist/issues/404 run TWEENs for the selected screen only
@@ -299,7 +445,7 @@
       if ( phet.chipper.queryParameters.supportsPanAndZoom ) {
 
         // animate the PanZoomListener, for smooth panning/scaling
-        animatedPanZoomSingleton.listener.step( dt );
+        animatedPanZoomSingleton.listener!.step( dt );
       }
 
       // View step is the last thing before updateDisplay(), so we can do paint updates there.
@@ -348,21 +494,14 @@
       }
     );
 
-    // @public (read-only) {HomeScreen|null}
     this.homeScreen = screenData.homeScreen;
-
-    // @public (read-only) {Screen[]} - the ordered list of sim-specific screens that appear in this runtime of the sim
     this.simScreens = screenData.selectedSimScreens;
 
-    // @public (read-only) {Screen[]} - all screens that appear in the runtime of this sim, with the homeScreen first if
-    // it was created
     this.screens = screenData.screens;
 
-    // @public (read-only) {boolean} - true if all possible screens are present (order-independent)
     this.allScreensCreated = screenData.allScreensCreated;
 
-    // @public {Property.<Screen>}
-    this.screenProperty = new Property( screenData.initialScreen, {
+    this.screenProperty = new Property<Screen<IntentionalAny, IntentionalAny>>( screenData.initialScreen, {
       tandem: Tandem.GENERAL_MODEL.createTandem( 'screenProperty' ),
       phetioFeatured: true,
       phetioDocumentation: 'Determines which screen is selected in the simulation',
@@ -370,8 +509,7 @@
       phetioType: Property.PropertyIO( Screen.ScreenIO )
     } );
 
-    // @public - the displayed name in the sim. This depends on what screens are shown this runtime (effected by query parameters).
-    this.displayedSimNameProperty = new DerivedProperty( [ this.simNameProperty, this.simScreens[ 0 ].nameProperty ],
+    this.displayedSimNameProperty = new DerivedProperty<string, [ string, string | null ]>( [ this.simNameProperty, this.simScreens[ 0 ].nameProperty ],
       ( simName, screenName ) => {
         const isMultiScreenSimDisplayingSingleScreen = this.simScreens.length === 1 && allSimScreens.length !== this.simScreens.length;
 
@@ -393,8 +531,6 @@
         }
       } );
 
-    // @public - When the sim is active, scenery processes inputs and stepSimulation(dt) runs from the system clock.
-    // Set to false for when the sim will be paused.
     this.activeProperty = new BooleanProperty( true, {
       tandem: Tandem.GENERAL_MODEL.createTandem( 'activeProperty' ),
       phetioFeatured: true,
@@ -402,7 +538,6 @@
                            'Setting this property to false pauses the simulation, and prevents user interaction.'
     } );
 
-    // @public (read-only) - property that indicates whether the browser tab containing the simulation is currently visible
     this.browserTabVisibleProperty = new BooleanProperty( true, {
       tandem: Tandem.GENERAL_MODEL.createTandem( 'browserTabVisibleProperty' ),
       phetioDocumentation: 'Indicates whether the browser tab containing the simulation is currently visible',
@@ -415,45 +550,26 @@
       this.browserTabVisibleProperty.set( document.visibilityState === 'visible' );
     }, false );
 
-    // @public (joist-internal, read-only) - How the home screen and navbar are scaled. This scale is based on the
-    // HomeScreen's layout bounds to support a consistently sized nav bar and menu. If this scale was based on the
-    // layout bounds of the current screen, there could be differences in the nav bar across screens.
     this.scaleProperty = new NumberProperty( 1 );
 
-    // @public (joist-internal, read-only) {Property.<Bounds2>|null} - global bounds for the entire simulation. null
-    //                                                                 before first resize
-    this.boundsProperty = new Property( null );
-
-    // @public (joist-internal, read-only) {Property.<Bounds2>|null} - global bounds for the screen-specific part
-    //                                                            (excludes the navigation bar), null before first resize
-    this.screenBoundsProperty = new Property( null );
-
-    // @public
+    this.boundsProperty = new Property<Bounds2 | null>( null );
+    this.screenBoundsProperty = new Property<Bounds2 | null>( null );
     this.lookAndFeel = new LookAndFeel();
     assert && assert( window.phet.joist.launchCalled, 'Sim must be launched using simLauncher, ' +
                                                       'see https://github.com/phetsims/joist/issues/142' );
 
-    // @private
     this.destroyed = false;
 
-    // @public {MemoryMonitor}
     this.memoryMonitor = new MemoryMonitor();
 
-    // public (read-only) {boolean} - if true, add support specific to accessible technology that work with touch devices.
     this.supportsGestureDescription = phet.chipper.queryParameters.supportsInteractiveDescription && SUPPORTS_GESTURE_DESCRIPTION;
-
-    // @public (joist-internal, read-only)
     this.hasKeyboardHelpContent = options.hasKeyboardHelpContent;
 
     assert && assert( !window.phet.joist.sim, 'Only supports one sim at a time' );
     window.phet.joist.sim = this;
 
-    // @public (read-only) {Property.<boolean>} - if PhET-iO is currently setting the state of the simulation.
-    // See PhetioStateEngine for details. This must be declared before soundManager.initialized is called.
     this.isSettingPhetioStateProperty = Tandem.PHET_IO_ENABLED ?
-                                        new DerivedProperty(
-                                          [ phet.phetio.phetioEngine.phetioStateEngine.isSettingStateProperty ],
-                                          _.identity ) :
+                                        phet.phetio.phetioEngine.phetioStateEngine.isSettingStateProperty :
                                         new BooleanProperty( false );
 
     // commented out because https://github.com/phetsims/joist/issues/553 is deferred for after GQIO-oneone
@@ -477,16 +593,11 @@
     // Make ScreenshotGenerator available globally so it can be used in preload files such as PhET-iO.
     window.phet.joist.ScreenshotGenerator = ScreenshotGenerator;
 
-    this.version = packageJSON.version; // @public (joist-internal)
-    this.credits = options.credits; // @public (joist-internal)
+    this.version = packageJSON.version;
+    this.credits = options.credits;
 
-    // @private - number of animation frames that have occurred
     this.frameCounter = 0;
-
-    // @private {boolean} - Whether the window has resized since our last updateDisplay()
     this.resizePending = true;
-
-    // @public - Make our locale available
     this.locale = phet.chipper.locale || 'en';
 
     // If the locale query parameter was specified, then we may be running the all.html file, so adjust the title.
@@ -495,12 +606,7 @@
       $( 'title' ).html( name );
     }
 
-    // @public {PreferencesManager} - If Preferences are available through a PreferencesConfiguration,
-    // this type will be added to the Sim to manage the state of features that can be enabled/disabled
-    // through user preferences.
     this.preferencesManager = null;
-
-    // @private {Toolbar|null} - The Toolbar is not created unless requested with a PreferencesConfiguration.
     this.toolbar = null;
 
     if ( options.preferencesConfiguration ) {
@@ -516,19 +622,16 @@
 
       // when the Toolbar positions update, resize the sim to fit in the available space
       this.toolbar.rightPositionProperty.lazyLink( () => {
-        this.resize( this.boundsProperty.value.width, this.boundsProperty.value.height );
+        this.resize( this.boundsProperty.value!.width, this.boundsProperty.value!.height );
       } );
     }
 
-    // @private
     this.display = new SimDisplay( options.simDisplayOptions );
-
-    // @public - root node for the Display
     this.rootNode = this.display.rootNode;
 
     Helper.initialize( this, this.display );
 
-    Property.multilink( [ this.activeProperty, phet.joist.playbackModeEnabledProperty ], ( active, playbackModeEnabled ) => {
+    Property.multilink( [ this.activeProperty, phet.joist.playbackModeEnabledProperty ], ( active: boolean, playbackModeEnabled: boolean ) => {
 
       // If in playbackMode is enabled, then the display must be interactive to support PDOM event listeners during
       // playback (which often come directly from sim code and not from user input).
@@ -548,10 +651,8 @@
 
     Heartbeat.start( this );
 
-    // @public (joist-internal)
     this.navigationBar = new NavigationBar( this, Tandem.GENERAL_VIEW.createTandem( 'navigationBar' ) );
 
-    // @public (joist-internal)
     this.updateBackground = () => {
       this.lookAndFeel.backgroundColorProperty.value = this.screenProperty.value.backgroundColorProperty.value;
     };
@@ -566,7 +667,6 @@
     // See https://github.com/phetsims/scenery/issues/218
     this.screenProperty.lazyLink( ( newScreen, oldScreen ) => oldScreen.view.interruptSubtreeInput() );
 
-    // @public (read-only) {SimInfo} - create this only after all other members have been set on Sim
     this.simInfo = new SimInfo( this );
 
     // Set up PhET-iO, must be done after phet.joist.sim is assigned
@@ -583,6 +683,22 @@
       }
     } );
 
+    this.topLayer = new Node();
+    this.modalNodeStack = createObservableArray(); // {Node} with node.hide()
+    this.barrierRectangle = new BarrierRectangle(
+      this.modalNodeStack, {
+        fill: 'rgba(0,0,0,0.3)',
+        pickable: true,
+        tandem: Tandem.GENERAL_VIEW.createTandem( 'barrierRectangle' ),
+        phetioDocumentation: 'Semi-transparent barrier used to block input events when a dialog is shown, also fades out the background'
+      } );
+    this.topLayer.addChild( this.barrierRectangle );
+
+    this.lastStepTime = -1;
+    this.lastAnimationFrameTime = -1;
+
+    this.boundRunAnimationLoop = this.runAnimationLoop.bind( this );
+
     // Third party support
     phet.chipper.queryParameters.legendsOfLearning && new LegendsOfLearningSupport( this ).start();
   }
@@ -609,11 +725,7 @@
     animationFrameTimer.runOnNextTick( () => phet.joist.display.updateDisplay() );
   }
 
-  /**
-   * @param {Screen[]} screens
-   * @private
-   */
-  finishInit( screens ) {
+  private finishInit( screens: Screen<IntentionalAny, IntentionalAny>[] ) {
 
     _.each( screens, screen => {
       screen.view.layerSplit = true;
@@ -625,7 +737,7 @@
       this.display.simulationRoot.addChild( this.toolbar );
       this.display.simulationRoot.pdomOrder = [ this.toolbar ];
 
-      // If Voicing is not "fully" enabled, only the toolbar is able to produce any Voicing output.
+      // If Voicing is not "fully" enabled, only the toolbar is able to produce Voicing output.
       // All other simulation components should not voice anything. This must be called only after
       // all ScreenViews have been constructed.
       voicingManager.voicingFullyEnabledProperty.link( fullyEnabled => {
@@ -653,29 +765,12 @@
 
         // Zoom out again after changing screens so we don't pan to the center of the focused ScreenView,
         // and so user has an overview of the new screen, see https://github.com/phetsims/joist/issues/682.
-        animatedPanZoomSingleton.listener.resetTransform();
+        animatedPanZoomSingleton.listener!.resetTransform();
       }
     } );
 
-    // layer for popups, dialogs, and their backgrounds and barriers
-    this.topLayer = new Node();
     this.display.simulationRoot.addChild( this.topLayer );
 
-    // @private list of nodes that are "modal" and hence block input with the barrierRectangle.  Used by modal dialogs
-    // and the PhetMenu
-    this.modalNodeStack = createObservableArray(); // {Node} with node.hide()
-
-    // @public (joist-internal) Semi-transparent black barrier used to block input events when a dialog (or other popup)
-    // is present, and fade out the background.
-    this.barrierRectangle = new BarrierRectangle(
-      this.modalNodeStack, {
-        fill: 'rgba(0,0,0,0.3)',
-        pickable: true,
-        tandem: Tandem.GENERAL_VIEW.createTandem( 'barrierRectangle' ),
-        phetioDocumentation: 'Semi-transparent barrier used to block input events when a dialog is shown, also fades out the background'
-      } );
-    this.topLayer.addChild( this.barrierRectangle );
-
     // Fit to the window and render the initial scene
     // Can't synchronously do this in Firefox, see https://github.com/phetsims/vegas/issues/55 and
     // https://bugzilla.mozilla.org/show_bug.cgi?id=840412.
@@ -696,15 +791,7 @@
     // Kick off checking for updates, if that is enabled
     updateCheck.check();
 
-    // @private - Keep track of the previous time for computing dt, and initially signify that time hasn't been recorded yet.
-    this.lastStepTime = -1;
-    this.lastAnimationFrameTime = -1;
-
-    // @public (joist-internal)
-    // Bind the animation loop so it can be called from requestAnimationFrame with the right this.
-    this.boundRunAnimationLoop = this.runAnimationLoop.bind( this );
-
-    // show any query parameter warnings in a dialog
+    // If there are warnings, show them in a dialog
     if ( QueryStringMachine.warnings.length ) {
       const warningDialog = new QueryParametersWarningDialog( QueryStringMachine.warnings, {
         closeButtonListener: () => {
@@ -720,11 +807,12 @@
    * Adds a popup in the global coordinate frame. If the popup is model, it displays a semi-transparent black input
    * barrier behind it. A modal popup prevent the user from interacting with the reset of the application until the
    * popup is hidden. Use hidePopup() to hide the popup.
-   * @param {Node} popup - the popup, must implemented node.hide(), called by hidePopup
-   * @param {boolean} isModal - whether popup is modal
-   * @public
+   * @param popup - the popup, must implemented node.hide(), called by hidePopup
+   * @param isModal - whether popup is modal
    */
-  showPopup( popup, isModal ) {
+
+  // TODO: https://github.com/phetsims/joist/issues/776 Better type for Popupable
+  showPopup( popup: Node & { hide: () => void; layout?: ( bounds: Bounds2 ) => void }, isModal: boolean ): void {
     assert && assert( popup );
     assert && assert( !!popup.hide, 'Missing popup.hide() for showPopup' );
     assert && assert( !this.topLayer.hasChild( popup ), 'popup already shown' );
@@ -733,18 +821,17 @@
       this.modalNodeStack.push( popup );
     }
     if ( popup.layout ) {
-      popup.layout( this.screenBoundsProperty.value );
+      popup.layout( this.screenBoundsProperty.value! );
     }
     this.topLayer.addChild( popup );
   }
 
   /*
    * Hides a popup that was previously displayed with showPopup()
-   * @param {Node} popup
-   * @param {boolean} isModal - whether popup is modal
-   * @public
+   * @param popup
+   * @param isModal - whether popup is modal
    */
-  hidePopup( popup, isModal ) {
+  hidePopup( popup: Node, isModal: boolean ): void {
     assert && assert( popup && this.modalNodeStack.includes( popup ) );
     assert && assert( this.topLayer.hasChild( popup ), 'popup was not shown' );
     if ( isModal ) {
@@ -753,28 +840,20 @@
     this.topLayer.removeChild( popup );
   }
 
-  /**
-   * @public (joist-internal)
-   */
-  resizeToWindow() {
+  private resizeToWindow(): void {
     this.resizePending = false;
     this.resize( window.innerWidth, window.innerHeight ); // eslint-disable-line bad-sim-text
   }
 
-  // @public (joist-internal, phet-io)
-  resize( width, height ) {
+  private resize( width: number, height: number ): void {
     this.resizeAction.execute( width, height );
   }
 
-  // Destroy a sim so that it will no longer consume any resources. Formerly used in Smorgasbord.  May not be used by
-  // anything else at the moment.
-
-  // @public (joist-internal)
-  start() {
+  start(): void {
 
     // In order to animate the loading progress bar, we must schedule work with setTimeout
     // This array of {function} is the work that must be completed to launch the sim.
-    const workItems = [];
+    const workItems: Array<() => void> = [];
 
     // Schedule instantiation of the screens
     this.screens.forEach( screen => {
@@ -792,7 +871,7 @@
     } );
 
     // loop to run startup items asynchronously so the DOM can be updated to show animation on the progress bar
-    const runItem = i => {
+    const runItem = ( i: number ) => {
       setTimeout( // eslint-disable-line bad-sim-text
         () => {
           workItems[ i ]();
@@ -806,7 +885,7 @@
           if ( document.getElementById( 'progressBarForeground' ) ) {
 
             // Grow the progress bar foreground to the right based on the progress so far.
-            document.getElementById( 'progressBarForeground' ).setAttribute( 'width', `${progress * PROGRESS_BAR_WIDTH}` );
+            document.getElementById( 'progressBarForeground' )!.setAttribute( 'width', `${progress * PROGRESS_BAR_WIDTH}` );
           }
           if ( i + 1 < workItems.length ) {
             runItem( i + 1 );
@@ -889,14 +968,16 @@
     runItem( 0 );
   }
 
-  // @public (joist-internal)
-  destroy() {
+  // Destroy a sim so that it will no longer consume resources. Formerly used in Smorgasbord.  May not be used by
+  // anything else at the moment.
+  // TODO: https://github.com/phetsims/joist/issues/776 remove this unused method
+  private destroy(): void {
     this.destroyed = true;
     this.display.domElement.parentNode && this.display.domElement.parentNode.removeChild( this.display.domElement );
   }
 
-  // @private - Bound to this.boundRunAnimationLoop so it can be run in window.requestAnimationFrame
-  runAnimationLoop() {
+  // Bound to this.boundRunAnimationLoop so it can be run in window.requestAnimationFrame
+  private runAnimationLoop(): void {
     if ( !this.destroyed ) {
       window.requestAnimationFrame( this.boundRunAnimationLoop );
     }
@@ -924,8 +1005,8 @@
     }
   }
 
-  // @private - run a single frame including model, view and display updates
-  stepOneFrame() {
+  // Run a single frame including model, view and display updates
+  stepOneFrame(): void {
 
     // Compute the elapsed time since the last frame, or guess 1/60th of a second if it is the first frame
     const currentTime = Date.now();
@@ -940,10 +1021,9 @@
 
   /**
    * Update the simulation model, view, scenery display with an elapsed time of dt.
-   * @param {number} dt in seconds
-   * @public (phet-io)
+   * @param dt - in seconds
    */
-  stepSimulation( dt ) {
+  stepSimulation( dt: number ): void {
     this.stepSimulationAction.execute( dt );
   }
 
@@ -951,10 +1031,8 @@
    * Hide or show all accessible content related to the sim ScreenViews, and navigation bar. This content will
    * remain visible, but not be tab navigable or readable with a screen reader. This is generally useful when
    * displaying a pop up or modal dialog.
-   * @param {boolean} visible
-   * @public
    */
-  setAccessibleViewsVisible( visible ) {
+  setAccessibleViewsVisible( visible: boolean ): void {
     for ( let i = 0; i < this.screens.length; i++ ) {
       this.screens[ i ].view.pdomVisible = visible;
     }
@@ -969,11 +1047,8 @@
    * and its buttons will be able to announce Voicing utterances. This is used by the
    * "Sim Voicing" switch in the toolbar which will disable all Voicing in the sim so that
    * only Toolbar content is announced.
-   * @public
-   *
-   * @param {boolean} visible
    */
-  setSimVoicingVisible( visible ) {
+  setSimVoicingVisible( visible: boolean ): void {
     for ( let i = 0; i < this.screens.length; i++ ) {
       this.screens[ i ].view.voicingVisible = visible;
     }
@@ -986,16 +1061,15 @@
 
 /**
  * Compute the dt since the last event
- * @param {number} lastTime - milliseconds, time of the last event
- * @param {number} currentTime - milliseconds, current time.  Passed in instead of computed so there is no "slack" between measurements
- * @returns {number} - seconds
+ * @param lastTime - milliseconds, time of the last event
+ * @param currentTime - milliseconds, current time.  Passed in instead of computed so there is no "slack" between measurements
+ * @returns - seconds
  */
-function getDT( lastTime, currentTime ) {
+function getDT( lastTime: number, currentTime: number ): number {
 
   // Compute the elapsed time since the last frame, or guess 1/60th of a second if it is the first frame
   return ( lastTime === -1 ) ? 1 / 60 :
          ( currentTime - lastTime ) / 1000.0;
 }
 
-joist.register( 'Sim', Sim );
-export default Sim;
\ No newline at end of file
+joist.register( 'Sim', Sim );
\ No newline at end of file
Index: main/joist/tsconfig.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/joist/tsconfig.json b/main/joist/tsconfig.json
--- a/main/joist/tsconfig.json	(revision e990d9da4f1f89581564ef6192b6466e6e433fcf)
+++ b/main/joist/tsconfig.json	(date 1649822777871)
@@ -9,6 +9,7 @@
     "../chipper/phet-types.d.ts",
     "../chipper/node_modules/@types/lodash/index.d.ts",
     "../chipper/node_modules/@types/qunit/index.d.ts",
+    "../chipper/node_modules/@types/jquery/index.d.ts",
 
     "js/**/*tests.js", // Cannot use **test** like we do elsewhere because of UpdaTESTate
     "js/**/*Tests.js", // Cannot use **test** like we do elsewhere because of UpdaTESTate
Index: main/joist/js/LookAndFeel.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/joist/js/LookAndFeel.ts b/main/joist/js/LookAndFeel.ts
--- a/main/joist/js/LookAndFeel.ts	(revision e990d9da4f1f89581564ef6192b6466e6e433fcf)
+++ b/main/joist/js/LookAndFeel.ts	(date 1649823911693)
@@ -1,5 +1,4 @@
 // Copyright 2015-2021, University of Colorado Boulder
-// @ts-nocheck
 /**
  * Provides colors for Joist elements.
  *
@@ -13,28 +12,35 @@
 
 class LookAndFeel {
 
+  // background color for the currently selected screen, which will be set on the Display as its backgroundColor
+  readonly backgroundColorProperty: Property<Color | string>;
+
+  // (joist-internal) True if the navigation bar background is black
+  readonly navigationBarDarkProperty: DerivedProperty<boolean, [ string | Color ]>;
+
+  // (joist-internal) - Navigation bar background fill
+  readonly navigationBarFillProperty: DerivedProperty<'white' | 'black', [ backgroundDark: boolean ]>;
+
+  // (joist-internal) - Navigation bar text fill
+  navigationBarTextFillProperty: DerivedProperty<'white' | 'black', [ navigationBarFill: 'white' | 'black' ]>;
+
   constructor() {
 
-    // @public background color for the currently selected screen, which will be set on the Display as its backgroundColor
-    this.backgroundColorProperty = new Property( 'black' );
+    this.backgroundColorProperty = new Property<Color | string>( 'black' );
 
-    // @public (joist-internal) {boolean} - True if the navigation bar background is black
-    this.navigationBarDarkProperty = new DerivedProperty( [ this.backgroundColorProperty ],
-      ( backgroundColor => {
-        return !!new Color( backgroundColor ).equals( Color.BLACK );
-      } ) );
+    this.navigationBarDarkProperty = new DerivedProperty<boolean, [ Color | string ]>( [ this.backgroundColorProperty ],
 
-    // @public (joist-internal) - Navigation bar background fill
+      // @ts-ignore
+      backgroundColor => !!new Color( backgroundColor ).equals( Color.BLACK )
+    );
+
     this.navigationBarFillProperty = new DerivedProperty( [ this.navigationBarDarkProperty ],
-      ( backgroundDark => {
-        return backgroundDark ? 'white' : 'black';
-      } ) );
+      backgroundDark => backgroundDark ? 'white' : 'black'
+    );
 
-    // @public (joist-internal) - Navigation bar text fill
     this.navigationBarTextFillProperty = new DerivedProperty( [ this.navigationBarFillProperty ],
-      ( navigationBarFill => {
-        return navigationBarFill === 'black' ? 'white' : 'black';
-      } ) );
+      navigationBarFill => navigationBarFill === 'black' ? 'white' : 'black'
+    );
   }
 
   // @public
Index: main/joist/js/Screen.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/joist/js/Screen.ts b/main/joist/js/Screen.ts
--- a/main/joist/js/Screen.ts	(revision e990d9da4f1f89581564ef6192b6466e6e433fcf)
+++ b/main/joist/js/Screen.ts	(date 1649825636780)
@@ -51,7 +51,10 @@
 type ScreenSelfOptions = {
   name?: string | null;
   instrumentNameProperty?: true;
-  backgroundColorProperty?: Property<string> | Property<Color>;
+
+  // It would be preferable to support Property<Color | string> solely, but many subtypes are hardcoded to be Color only
+  // or string only, so we support this polymorphic form
+  backgroundColorProperty?: Property<Color | string> | Property<Color> | Property<string>;
   homeScreenIcon?: ScreenIcon | null;
   showUnselectedHomeScreenIconFrame?: boolean;
   navigationBarIcon?: ScreenIcon | null;
@@ -64,21 +67,23 @@
 
 // Parameterized on M=Model and V=View
 class Screen<M, V extends ScreenView> extends PhetioObject {
-  backgroundColorProperty: Property<string> | Property<Color>;
-  private readonly nameProperty: IReadOnlyProperty<string | null>;
+
+  backgroundColorProperty: Property<Color> | Property<string> | Property<Color | string>;
+  readonly maxDT: number;
+  readonly activeProperty: BooleanProperty;
+  readonly descriptionContent: string;
+  readonly nameProperty: IReadOnlyProperty<string | null>;
+
   private readonly showScreenIconFrameForNavigationBarFill: string | null;
   private readonly homeScreenIcon: Node | null;
   private readonly navigationBarIcon: Node | null;
   private readonly showUnselectedHomeScreenIconFrame: boolean;
   private readonly keyboardHelpNode: Node | null; // joist-internal
   private readonly pdomDisplayNameProperty: DerivedProperty<string | null, [ string | null ]>;
-  private readonly maxDT: number;
   private readonly createModel: () => M;
   private readonly createView: ( model: M ) => V;
   private _model: M | null;
   private _view: V | null;
-  private readonly activeProperty: BooleanProperty;
-  public readonly descriptionContent: string;
 
   static HOME_SCREEN_ICON_ASPECT_RATIO: number;
   static MINIMUM_HOME_SCREEN_ICON_SIZE: Dimension2;
@@ -97,8 +102,7 @@
       // {boolean} whether nameProperty should be instrumented. see usage for explanation of its necessity.
       instrumentNameProperty: true,
 
-      // {Property.<Color|string>} background color of the Screen
-      backgroundColorProperty: new Property( 'white' ),
+      backgroundColorProperty: new Property<Color | string>( 'white' ),
 
       // {Node|null} icon shown on the home screen. If null, then a default is created.
       // For single-screen sims, there is no home screen and the default is OK.
@@ -294,7 +298,7 @@
     }
 
     // Set the accessible label for the screen.
-    Property.multilink<[ string, string, string | null]>( [ displayedSimNameProperty, simNameProperty, this.pdomDisplayNameProperty ],
+    Property.multilink<[ string, string, string | null ]>( [ displayedSimNameProperty, simNameProperty, this.pdomDisplayNameProperty ],
       ( displayedName, simName, pdomDisplayName ) => {
 
         let titleString;
Index: main/joist/tsconfig-module.json
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/joist/tsconfig-module.json b/main/joist/tsconfig-module.json
--- a/main/joist/tsconfig-module.json	(revision e990d9da4f1f89581564ef6192b6466e6e433fcf)
+++ b/main/joist/tsconfig-module.json	(date 1649822837618)
@@ -17,6 +17,7 @@
     "../chipper/phet-types.d.ts",
     "../chipper/node_modules/@types/lodash/index.d.ts",
     "../chipper/node_modules/@types/qunit/index.d.ts",
+    "../chipper/node_modules/@types/jquery/index.d.ts",
 
     "js/**/*",
     "sounds/**/*",
Index: main/joist/js/SimDisplay.js
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/joist/js/SimDisplay.js b/main/joist/js/SimDisplay.js
--- a/main/joist/js/SimDisplay.js	(revision e990d9da4f1f89581564ef6192b6466e6e433fcf)
+++ b/main/joist/js/SimDisplay.js	(date 1649823140891)
@@ -166,7 +166,7 @@
 
   /**
    * Handle synthetic input event fuzzing
-   * @private
+   * @public
    */
   fuzzInputEvents() {
 
Index: main/joist/js/selectScreens.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/joist/js/selectScreens.ts b/main/joist/js/selectScreens.ts
--- a/main/joist/js/selectScreens.ts	(revision e990d9da4f1f89581564ef6192b6466e6e433fcf)
+++ b/main/joist/js/selectScreens.ts	(date 1649825636767)
@@ -1,6 +1,7 @@
 // Copyright 2020-2021, University of Colorado Boulder
 // @ts-nocheck
 import joist from './joist.js';
+import HomeScreen from './HomeScreen.js';
 
 /**
  * Given an array of all possible screens that a sim can have, select and order them according to the relevant query
@@ -13,25 +14,26 @@
  * @author Michael Kauzmann (PhET Interactive Simulations)
  * @author Sam Reid (PhET Interactive Simulations)
  *
- * @param {Screen[]} allSimScreens - all of the screens declared by the sim, duck-typed for tests
- * @param {boolean} homeScreenQueryParameter - from phet.chipper.queryParameters.homeScreen
- * @param {boolean} homeScreenQueryParameterProvided
- * @param {number} initialScreenIndex - from phet.chipper.queryParameters.initialScreen
- * @param {boolean} initialScreenQueryParameterProvided
- * @param {number[]} screensQueryParameter - from phet.chipper.queryParameters.screens
- * @param {boolean} screensQueryParameterProvided
- * @param {function( selectedSimScreens ):HomeScreen} createHomeScreen
+ * @param allSimScreens - all of the screens declared by the sim, duck-typed for tests
+ * @param homeScreenQueryParameter - from phet.chipper.queryParameters.homeScreen
+ * @param homeScreenQueryParameterProvided
+ * @param initialScreenIndex - from phet.chipper.queryParameters.initialScreen
+ * @param initialScreenQueryParameterProvided
+ * @param screensQueryParameter - from phet.chipper.queryParameters.screens
+ * @param screensQueryParameterProvided
+ * @param createHomeScreen
  * @returns {{homeScreen:HomeScreen|null, initialScreen:Screen, selectedSimScreens:Screen[], screens:Screen[]}} - duck-typed for tests
  * @throws Error if incompatible data is provided
  */
-const selectScreens = ( allSimScreens,
-                        homeScreenQueryParameter,
-                        homeScreenQueryParameterProvided,
-                        initialScreenIndex,
-                        initialScreenQueryParameterProvided,
-                        screensQueryParameter,
-                        screensQueryParameterProvided,
-                        createHomeScreen ) => {
+// TODO: Explicit return type, see https://github.com/phetsims/joist/issues/776
+const selectScreens = ( allSimScreens: Screen<any, any>[],
+                        homeScreenQueryParameter: boolean,
+                        homeScreenQueryParameterProvided: boolean,
+                        initialScreenIndex: number,
+                        initialScreenQueryParameterProvided: boolean,
+                        screensQueryParameter: number[],
+                        screensQueryParameterProvided: boolean,
+                        createHomeScreen: ( screens: Screen<any, any>[] ) => HomeScreen ) => {
 
   if ( allSimScreens.length === 1 && homeScreenQueryParameterProvided && homeScreenQueryParameter ) {
     const errorMessage = 'cannot specify homeScreen=true for single-screen sims';
Index: main/geometric-optics/js/GOSim.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/main/geometric-optics/js/GOSim.ts b/main/geometric-optics/js/GOSim.ts
--- a/main/geometric-optics/js/GOSim.ts	(revision 384d6b2b37584c5cce98162c0d8b120228e62069)
+++ b/main/geometric-optics/js/GOSim.ts	(date 1649825320980)
@@ -6,7 +6,7 @@
  * @author Chris Malley (PixelZoom, Inc.)
  */
 
-import Sim from '../../joist/js/Sim.js';
+import Sim, { SimOptions } from '../../joist/js/Sim.js';
 import Tandem from '../../tandem/js/Tandem.js';
 import geometricOptics from './geometricOptics.js';
 import LensScreen from './lens/LensScreen.js';
@@ -14,14 +14,6 @@
 import GOConstants from './common/GOConstants.js';
 import optionize from '../../phet-core/js/optionize.js';
 import GOOptionsNode from './common/view/GOOptionsNode.js';
-import { Node } from '../../scenery/js/imports.js';
-
-//TODO https://github.com/phetsims/joist/issues/795 import SimOptions from Sim.js
-type SimOptions = {
-  credits?: object;
-  hasKeyboardHelpContent?: boolean;
-  createOptionsDialogContent?: ( tandem: Tandem ) => Node;
-};
 
 type SelfOptions = {
 

```

</details>

**Comment by pixelzoom** [comment]
> I feel it has been rare for developers to work much before 8am ...

Not at all rare for me.  I'm often trying to bust out a few things so that they're ready for review by QA or my sim team.

**Comment by pixelzoom** [comment]
> I feel it has been rare for developers to work much before 8am ...

Not at all rare for me.  I'm often trying to bust out a few things so that they're ready for review by QA or my sim team.  Some of the dev team is in ET, and  8am MT is 10am ET.

**Comment by pixelzoom** [comment]
It sounded like you didn't commit to master.  But CT is currently bright red.  So I'm not going to investigate, and will defer my morning work plans (again).

**Comment by samreid** [comment]
Sounds good.  I feel I'm still grappling with the ramifications of the recent discussion about trying to be around to monitor CT before pushing nontrivial changes.  I think I'll open a side issue for that aspect.

> It sounded like you didn't commit to master. But CT is currently bright red. So I'm not going to investigate, and will defer my morning work plans (again).

My fault, I'll push a revert. In this case it may have been better for me to push more elaborate changes to CT because the problem was caused by this:

> I committed ts-nocheck and Sim.js => ts before I knew the scope of this project.

**Comment by pixelzoom** [comment]
>  I feel I'm still grappling with the ramifications of the recent discussion about trying to be around to monitor CT before pushing nontrivial changes

What is the problem with simply waiting to push untl you have time to check CT?  I don't get it... And this continues to impact my abilty to work.

**Comment by pixelzoom** [comment]
And there was no need to "stick around" in this case.  Did you even check CT "quick view"? It would have immediately shown the problem. What good would a faster CT do if you don't even look at it?

**Comment by samreid** [comment]
I reverted the errant commits and CT should be clear.

**Comment by samreid** [comment]
I added a note to https://github.com/phetsims/phet-info/blob/master/doc/typescript-quick-start.md

> Please note that adding JS=>TS with a `// ts-nocheck` can introduce type checking errors in other files though, so this is not a foolproof strategy.

**Comment by samreid** [comment]
See also https://github.com/phetsims/joist/issues/795 where there was other work toward Sim JS=>TS.

**Comment by zepumph** [comment]
Home screen code done above.

**Comment by zepumph** [comment]
Lots of converting in joist today. Some questions:

* Should Screen params default to IModel and ScreenView? I think it will simplify a fair number of spots in joist.

**Comment by zepumph** [comment]
Ok. All files in Joist have been converted, and most ts-ignores were able to be deleted now that other common code has been brought to typescript. I think we are ready to close. Remaining work lives in other issues.
